Green thread objects

davidwrighton opened the pull request:
This adds managed `Thread` objects for green threads as well as unique managed thread ids. There is a test for that. 

In addition this adds logic to moves handling of Monitor locks to a per-green thread structure.

In addition it fixes a number of bugs

- In the reverse p/invoke logic where the current native `Thread*` pointer is cached, and when we hop threads isn't properly restored. (See change in jithelpers.cpp)
- This also fixes bugs where the stack limits were not properly updated on resume
- This fixes bugs where in the presence of optimization, when the thread was resumed on a different thread the Thread* was not refreshed from the underlying TLS data structure. This issue was found in several functions in C++ (which are now marked with optimizations disabled, and the GreenThread_Yield P/Invoke which has the same issue.)

Fixes #2014

davidwrighton committed:
Protect Thread object from being called on GreenThreads

davidwrighton committed:
Add concept of ThreadBase which holds the thread id and is used as part of synchronization

davidwrighton committed:
Move Thread ID and monitor lock infrastructure into a new data structure that can be per-green thread

davidwrighton committed:
Move exposed object out of Thread

davidwrighton committed:
Actually hook it all up. I wonder if it works?

davidwrighton committed:
Fix bug when the green thread hops from one OS thread to another

davidwrighton committed:
Fix bugs, and add test

davidwrighton committed:
Fix release build break

davidwrighton committed:
Fix test build failure

cshung reviewed (approved):
I read through the changes in Thread/SyncBlock, I only skimmed through the refactoring.
Looks good to me.

davidwrighton committed:
Fix a swath of bugs causing optimized builds to fail
1. TEB wasn't properly updated on resume
2. The Thread pointer is cached across a green thread suspension/resume point in a few places. Mitigate by disabling optimization or by disabling fully inlined PInvokes. This is proving to be tricky to keep right.

davidwrighton committed:
Fix more memory corruption issues and add a new contended lock test

davidwrighton committed:
Fix issue with handling of IL stubs and pinvoke helper functions instead of fully inlined pinvokes

davidwrighton committed:
Merge branch 'feature/green-threads' of github.com:dotnet/runtimelab into GreenThreadObjects

davidwrighton committed:
Remove ReadLine used to help debugging
